enhancements/cluster-high-availability-mode-api.md

```diff
@@ -0,0 +1,360 @@
+supported in the older version.
+
+### Version Skew Strategy
+
+Clusters deployed using 4.7 or earlier will not have the new field in
+the `infrastructure` API so the value in the struct will be
+empty. Operators that need the high-availability mode should treat an
+empty string as `Full` high-availability mode.
+
+## Implementation History
+
+Major milestones in the life cycle of a proposal should be tracked in `Implementation
+History`.
+
+## Drawbacks
+
+The idea is to find the best form of an argument why this enhancement should _not_ be implemented.
+
+## Alternatives
+
+### Define a new Capabilities API
+
+The initial proposal was to create a new informational API to express
+cluster-wide "capabilities". That approach was rejected for several
+reasons:
+
+1. In configurations where there is an external control plane, the
+   `infrastructure` API is already being expressed inside the
+   cluster. Adding a new API would require similar work.
+2. The new API was seen as collecting unrelated settings into a
+   generic object.
+
+In addition to these objections, a new API is more work than adding a
+field (or a few fields) to the existing API, because it requires more
+code to be updated, including having the RBAC rules configured to
+allow the service accounts to see it.
+
+### Expose the cluster profile name directly
+
+We could add a `clusterProfile` field to an existing or new
+configuration API. This would require each operator to interpret each
+new profile, which would make maintaining profiles even more work than
+it is now. It may also encourage us to create more profiles, when we
+have stated elsewhere that our goal is to have as few as possible.
+
+### Use the cluster profile to have different Deployments for operators
+
+We could use different manifests to have the cluster-version-operator
+deploy operators with different configuration (via command line
+arguments or environment variables). This would give us a way to
+configure operators differently for each profile, but those settings
+would be less discoverable and we would not be able to offer users a
+way to change them at runtime (if we ever want to do that).
+
+### Recognize HA/non-HA configuration based on control plane node count
+
+Operators could look at the number of control plane nodes and use that
+information to determine whether HA is enabled (3 or more nodes) or
+not (1 node).
+
+This would require giving the operator read access to Node resources,
+and would be subject to race conditions as the control plane
+forms. For example, the cluster-etcd-operator [requires 3 control
+plane
+nodes](https://github.com/openshift/cluster-etcd-operator/blob/98590e6ecfe282735c4eff01432ae40b29f81202/pkg/etcdenvvar/etcd_env.go#L72)
+to exist before deploying its operand. If it counted nodes, it could
+not tell the difference between a cluster starting to form with 1
+node, then more, or a cluster that is only ever going to include 1
+node.
```

Comment:
I think the key question here is whether the amount of masters is orthogonal to the desired state.  Given that we don't want a cluster of size 3 accidentally reduced to 1 to flip into non-HA mode, we should not use "number of masters" to indicate the setting.  We could maybe come up with behaviors that say "revert to being quorum of 1 if you only have one node", but we don't *want* those in failure scenarios.

Reply:
See also the discussion of why the input replica count does not cover all cases in the alternative described on lines 288-306.

Reply:
>  Given that we don't want a cluster of size 3 accidentally reduced to 1 to flip into non-HA mode

If we can not describe what HA is in the cluster topology during runtime (masters/machines/X) how do we know if the cluster is actually HA? Today operators need to count master nodes. I would expect if the HA mode could not be enforced some infra controller would be degraded communicating this to the admin/cluster. Otherwise, operators will need to use node counts to go degraded themselves.

I am operator HA Mode is `None` master node count is 2 who is degraded? It's these corner cases that make me nervous.

Reply:
Should etcd operator own degraded status in this case?

Reply:
Does an operator have a way of discovering how many nodes a control plane is intended to have?

Reply:
> Does an operator have a way of discovering how many nodes a control plane is intended to have?

There isn't anything today or in this proposal that would allow that. Do operators need to know the actual number?

Reply:
> Do operators need to know the actual number?
Some do specifically etcd-operator is concerned with the definition of HA. Our definition of HA is a guarantee that the cluster can provide fault tolerance, which allows for rolling reboots of nodes without quorum loss. The guarantee is directly related to the number of control-plane nodes today. less than 3 HA `None`.

Reply:
> > Do operators need to know the actual number?
> 
> Some do specifically etcd-operator is concerned with the definition of HA. Our definition of HA is a guarantee that the cluster can provide fault tolerance, which allows for rolling reboots of nodes without quorum loss. The guarantee is directly related to the number of control-plane nodes today. less than 3 HA `None`.

Is the actual variable driving etcd HA properties the number of control plane nodes, or the number of etcd members, which happen to be (in the stock OCP implementation) coupled to control plane nodes?

Tangentially, other implementations where HA etcd can be achieved without such coupling: containerized etcd, externalized etcd.

Reply:
Would it be useful to have a separate discussion elsewhere about the perceived requirement that the etcd-operator be aware of control plane replica counts? The assumption seems to be that replica awareness is essential, which could be worth challenging with fresh eyes just in case it's really not

Reply:
Yes, I think the discussion of operators needing to know the actual expected replica count should be separate from this change. For the purposes of this enhancement, we only need to be able to differentiate between 1 and 3. Four shalt thou not count, neither count thou two, excepting that thou then proceed to three.

Reply:
> Would it be useful to have a separate discussion elsewhere about the perceived requirement that the etcd-operator be aware of control plane replica counts? The assumption seems to be that replica awareness is essential, which could be worth challenging with fresh eyes just in case it's really not

I look forward to this discussion.